---
title: "Initial GTEX Analysis for Lung"
author: David Gerard
output: html_document
date: "`r Sys.Date()`"
bibliography: gtex_bib.bib
---

# Brief Summary
Read in the cleaned lung gtex data and estimate the number of hidden confounders with `sva::num.sv()`.
```{r, cache = TRUE}
library(ggplot2)
load(file = "../output/cleaned_gtex_data/lung.Rd")
k <- sva::num.sv(dat = lung$Y, mod = lung$X)
k
```

$Y$ is log(COUNTS + 1) of the top `r nrow(lung$Y)` expressed genes (genes with at least 10 reads per sample on average). Samples from the same individual were averaged, resulting in $n =$ `r ncol(lung$Y)`. $X$ contains just an intercept and an indicator for gender. The idea here is that genes on sex chromosomes are "positive controls," so the best method will be that which has the most genes on sex chromosomes among the most significant genes. This is the setup that @gagnon2012using and @wang2015confounder use to evaluate performance between these confounder adjustment methods.

Control genes are the housekeeping genes identified by @eisenberg2013human. This is a different list than that used by @gagnon2012using, which was drawn from @eisenberg2003human. In total, there are `r sum(lung$ctl)` housekeeping genes among the `r nrow(lung$Y)` most expressed genes in the GTEX lung data.

# Analysis

## RUVASH vs RUV4

Now fit RUVASH and RUV4.
```{r, cache = TRUE}
qthresh <- 0.05
onsex <- lung$chrom == "X" | lung$chrom == "Y"
ruvash_out <- ashr::ash_ruv(Y = t(lung$Y), X = lung$X, ctl = lung$ctl, k = k)
ruvash_onsex <- onsex[order(ruvash_out$lfdr)]
ruv4_out <- ruv::RUV4(Y = t(lung$Y), X = lung$X[, 2, drop = FALSE], ctl = lung$ctl, k = k)
ruv4_onsex <- onsex[order(ruv4_out$p)]
num_sig <- sum(ruvash_out$qvalue < qthresh)
qout <- qvalue::qvalue(p = ruv4_out$p)
num_sig_ruv4 <- sum(qout$qvalue < qthresh)
```

The orderings that RUVASH and RUV4 give for the most significant genes are very similar. RUVASH says that `r num_sig` genes have a q-value less than `r qthresh`. RUVASH says that `r sum(ruvash_onsex[1:num_sig])` of the top `r num_sig` genes are on a sex chromosome while RUV4 says that `r sum(ruv4_onsex[1:num_sig])` of the top `r num_sig` genes are on a sex chromosome. We can calculate the Kendall's tau of the rankings that RUV4 and RUVASH give for the top `r num_sig` genes:
```{r}
lfdr_order <- order(ruvash_out$lfdr)
tau <- cor(c(ruv4_out$p)[lfdr_order[1:num_sig]], ruvash_out$lfdr[lfdr_order[1:num_sig]], method = "kendall")
tau
```

Their rankings don't differ that much in the first `r num_sig` genes. The ordering later on gets a very different.
```{r}
qplot(rank(c(ruv4_out$p))[lfdr_order[1:num_sig]], rank(ruvash_out$lfdr)[lfdr_order[1:num_sig]],
      xlab = "RUV4 Rank", ylab = "RUVASH Rank",
      main = paste("Rankings for top", num_sig, "genes"),
      color = onsex[lfdr_order[1:num_sig]]) +
    guides(color=guide_legend(title="On Sex?"))

qplot(rank(c(ruv4_out$p)), rank(ruvash_out$lfdr),
      xlab = "RUV4 Rank", ylab = "RUVASH Rank",
      main = "Rankings for all genes",
      color = onsex[lfdr_order]) +
    guides(color=guide_legend(title="On Sex?"))
```
<!-- With a Kendall's tau of `r cor(c(ruv4_out$p), ruvash_out$lfdr, method = "kendall")`. -->

However, I am going to argue that RUVASH is much better calibrated than RUV4. RUV4, followed by using `qvalue` [@storey2003positive], estimates $\pi_0$ to be `r qout$pi0` against RUVASH's estimated $\pi_0$ of `r ruvash_out$fitted.g$pi[1]`. I think RUVASH's estimate is much more realistic in this scenario as `r 1 - mean(onsex)` of the genes are not on a sex chromosome, so one would think that $\pi_0$ is closer to this value. Furthermore, RUV4 + `qvalue` says that `r num_sig_ruv4` genes have a q-value less than `r qthresh`, compared to RUVASH's estimate of `r num_sig`. The proportion of the top `r num_sig` RUVASH genes on a sex chromosome is `r mean(ruvash_onsex[1:num_sig])`. The proportion of the top `r num_sig_ruv4` RUV4 genes on a sex chromosome is `r mean(ruv4_onsex[1:num_sig_ruv4])`, which seems like very poor performance.

The shift from sex chromosome genes to non-sex chromosome genes is not as clear as in the heart data set.
```{r}
num_plot <- 200
df_lfdr <- data.frame(lfdr = ruvash_out$lfdr[lfdr_order][1:num_plot],
                      qvalue = ruvash_out$qvalue[lfdr_order][1:num_plot],
                      onsex = ruvash_onsex[1:num_plot],
                      index = 1:num_plot)
ggplot(data = df_lfdr, mapping = aes(y = lfdr, x = index, color = onsex)) + geom_point() +
    geom_vline(xintercept = num_sig) +
    ggtitle(paste("lfdrs for top", num_plot, "genes"))

ggplot(data = df_lfdr, mapping = aes(x = qvalue, fill = onsex)) + geom_histogram(bins = 25) +
    geom_vline(xintercept = qthresh) +
    ggtitle(paste("Histogram of qvalues for top", num_plot, "genes"))
```

RUV4 is not having a problem with the p-value distribution. Indeed, it looks pretty good:
```{r}
ggplot(data = data.frame(pvalue = c(ruv4_out$p)), aes(x = pvalue, color = I("black"))) +
    geom_histogram(, bins = 20)
```

The t-statistics for the genes that RUVASH says is null match up pretty well against a standard normal.
```{r}
sig_genes <- ruvash_out$qvalue < 0.05
tstat <- ruvash_out$ruv$betahat / ruvash_out$ruv$sebetahat
qqnorm(tstat[!sig_genes])
qqline(tstat[!sig_genes])
```

## ASH using RUV4 without the variance inflation.
The variance inflation for RUVASH is very high, being `r ruvash_out$ruv$multiplier`. We can fit ASH using the output from RUV4 and see what comes out.

```{r, cache = TRUE}
betahat   <- ruv4_out$betahat
sebetahat <- sqrt(ruv4_out$sigma * ruv4_out$multiplier)
ash_out <- ashr::ash(betahat = betahat, sebetahat = sebetahat)
ash_out$fitted.g$pi[1] ## estimate of pi0
```
which is even lower than that provided by `qvalue`.

## RUV4, ASH and other choices of number of confounders.

It has been suggested that if I try different numbers of confounders, then RUV4 and ash might perform better. Rather, that the variance inflation is needed mostly because I am not estimating the number of hidden confounders accurately. If I run RUV4 with different choices of number of hidden confounders, I always get estimates of $\pi_0$ very far from what RUVASH is giving me. The RUVASH estimate seems more reasonable given the proportion of genes on sex chromosomes.
```{r, cache = TRUE}
maxk <- 50
pi0mat_q <- matrix(NA, nrow = maxk + 1, ncol = 3)
num_onsex_mat <- matrix(NA, nrow = maxk + 1, ncol = 3)
for (index in 0:maxk) {
    ruv4_out <- ruv::RUV4(Y = t(lung$Y), X = lung$X[, 2, drop = FALSE],
                          ctl = lung$ctl, k = index)
    qout <- qvalue::qvalue(ruv4_out$p)
    betahat   <- ruv4_out$betahat
    sebetahat <- sqrt(ruv4_out$sigma * ruv4_out$multiplier)
    ash_out <- ashr::ash(betahat = betahat, sebetahat = sebetahat)
    ash_onsex_iter <- onsex[order(ash_out$lfdr)]
    ruv4_onsex_iter <- onsex[order(ruv4_out$p)]
    pi0mat_q[index + 1, 1] <- qout$pi0
    pi0mat_q[index + 1, 2] <- ash_out$fitted.g$pi[1]
    pi0mat_q[index + 1, 3] <- index

    num_onsex_mat[index + 1, 1] <- sum(ruv4_onsex_iter[1:num_sig])
    num_onsex_mat[index + 1, 2] <- sum(ash_onsex_iter[1:num_sig])
    num_onsex_mat[index + 1, 3] <- index
}
```




```{r}
colnames(pi0mat_q) <- c("RUV4andQvalue", "RUV4andASH", "num_confounder")
longdat <- reshape2::melt(as.data.frame(pi0mat_q), id.vars = "num_confounder")
ggplot(data = longdat, aes(y = value, x = num_confounder, color = variable)) +
    geom_line() + ylab(expression(hat(pi)[0])) +
    xlab("Assumed Number of Hidden Confounders") +
    ggtitle("Estimated pi0 vs number of hidden confounders") +
    guides(color=guide_legend(title="Method")) +
    scale_color_discrete(labels = c("RUV4 + qvalue", "RUV4 + ASH"))
```

Recall that the `num.sv` estimated the number of hidden confounders to be `r k` and the estimate of $\pi_0$ given by RUVASH is `r ruvash_out$fitted.g$pi[1]`. They do pretty well at smaller numbers of hidden confounders, but the ordering does not do as well. It seems the optimal number of hidden confounders for RUV4 + ASH is around 15, but with very little robustness.


```{r}
colnames(num_onsex_mat) <- c("RUV4andQvalue", "RUV4andASH", "num_confounder")
longdat <- reshape2::melt(as.data.frame(num_onsex_mat), id.vars = "num_confounder")
ggplot(data = longdat, aes(y = value, x = num_confounder, color = variable)) +
    geom_line() + ylab("Counts") +
    xlab("Assumed Number of Hidden Confounders") +
    ggtitle(paste("Number of Sex Genes in First", num_sig, "Genes")) +
    guides(color=guide_legend(title="Method")) +
    scale_color_discrete(labels = c("RUV4 + qvalue", "RUV4 + ASH"))
```


## What About OLS + Limma?
```{r}
ols_out <- limma::lmFit(object = lung$Y, design = lung$X)
lim_out <- limma::ebayes(fit = ols_out)
lim_p <- lim_out$p.value[, 2]
lim_onsex <- onsex[order(lim_p)]
```

Limma says that `r sum(lim_onsex[1:num_sig])` of the top `r num_sig` are on a sex chromosome. The ordering of genes via Limma and RUVASH are very different.
```{r}
qplot(rank(c(lim_p))[lfdr_order[1:num_sig]], rank(ruvash_out$lfdr)[lfdr_order[1:num_sig]],
      xlab = "OLS Rank", ylab = "RUVASH Rank", main = paste("Rankings of first", num_sig, "genes"),
      color = onsex[lfdr_order[1:num_sig]]) +
    guides(color=guide_legend(title="On Sex?"))

qplot(rank(c(lim_p)), rank(ruvash_out$lfdr),
      xlab = "OLS Rank", ylab = "RUVASH Rank", main = paste("Rankings of all genes"),
      color = onsex) +
    guides(color=guide_legend(title="On Sex?"))
```


```{r}
sessionInfo()
```

# References
